datafusion/optimizer/src/type_coercion.rs

```diff
@@ -79,10 +79,29 @@ impl OptimizerRule for TypeCoercion {
             const_evaluator,
         };
 
+        let original_expr_names: Vec<Option<String>> = plan
+            .expressions()
+            .iter()
+            .map(|expr| expr.name().ok())
+            .collect();
+
         let new_expr = plan
             .expressions()
             .into_iter()
-            .map(|expr| expr.rewrite(&mut expr_rewrite))
+            .zip(original_expr_names)
+            .map(|(expr, original_name)| {
+                let expr = expr.rewrite(&mut expr_rewrite)?;
+
+                if matches!(expr, Expr::AggregateFunction { .. }) {
```

Comment:
```suggestion
                // ensure aggregate names don't change:
                // https://github.com/apache/arrow-datafusion/issues/3555
                if matches!(expr, Expr::AggregateFunction { .. }) {
```

Reply:
I think this is just a workround for this issue. We should split the type coercion from the optimizer

Reply:
Yeah, I agree